packages/tenant-process/src/services/tenantService.ts

```diff
@@ -0,0 +1,115 @@
+import { AuthData, eventRepository, initDB } from "pagopa-interop-commons";
+import { Tenant, tenantEventToBinaryData } from "pagopa-interop-models";
+import { v4 as uuidv4 } from "uuid";
+import { TenantProcessConfig } from "../utilities/config.js";
+import {
+  toCreateEventTenantAdded,
+  toCreateEventTenantUpdated,
+} from "../model/domain/toEvent.js";
+import { ApiSelfcareTenantSeed } from "../model/types.js";
+import { selfcareIdConflict } from "../model/domain/errors.js";
+import {
+  assertResourceAllowed,
+  getTenantKind,
+  getTenantKindLoadingCertifiedAttributes,
+} from "./validators.js";
+import { ReadModelService } from "./readModelService.js";
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function tenantServiceBuilder(
+  config: TenantProcessConfig,
+  readModelService: ReadModelService
+) {
+  const repository = eventRepository(
+    initDB({
+      username: config.eventStoreDbUsername,
+      password: config.eventStoreDbPassword,
+      host: config.eventStoreDbHost,
+      port: config.eventStoreDbPort,
+      database: config.eventStoreDbName,
+      schema: config.eventStoreDbSchema,
+      useSSL: config.eventStoreDbUseSSL,
+    }),
+    tenantEventToBinaryData
+  );
+
+  return {
+    async selfcareUpsertTenant({
+      tenantSeed,
+      authData,
+    }: {
+      tenantSeed: ApiSelfcareTenantSeed;
+      authData: AuthData;
+    }): Promise<string> {
+      const existingTenant = await readModelService.getTenantByExternalId(
+        tenantSeed.externalId
+      );
+      if (existingTenant) {
+        await assertResourceAllowed(existingTenant.data.id, authData);
+        const tenantKind = await getTenantKindLoadingCertifiedAttributes(
+          readModelService,
+          existingTenant.data.attributes,
+          existingTenant.data.externalId
+        );
+
+        const updatedSelfcareId = evaluateNewSelfcareId({
+          tenant: existingTenant.data,
+          newSelfcareId: tenantSeed.selfcareId,
+        });
```

Comment:
```suggestion

        const updatedSelfcareId = evaluateNewSelfcareId({
          tenant: existingTenant.data,
          newSelfcareId: tenantSeed.selfcareId,
        });

        const tenantKind = await getTenantKindLoadingCertifiedAttributes(
          readModelService,
          existingTenant.data.attributes,
          existingTenant.data.externalId
        );
```

It's better to swap the order of these two because if the evaluation fails there's no need to continue

Reply:
Interesting, thanks